libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -875,6 +875,61 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		m_code << templ.render();
 		break;
 	}
+	case FunctionType::Kind::ABIDecode:
+	{
+		Whiskers templ(R"(
+			let <retVars> := <abiDecode>(<offset>, add(<offset>, add(<offset>, <length>)))
+		)");
+
+		TypePointer firstArgType = arguments.front()->annotation().type;
+		TypePointers targetTypes;
+
+		if (TupleType const* targetTupleType = dynamic_cast<TupleType const*>(_functionCall.annotation().type))
+			targetTypes = targetTupleType->components();
+		else
+			targetTypes = TypePointers{_functionCall.annotation().type};
+
+		if (
+			auto referenceType = dynamic_cast<ReferenceType const*>(firstArgType);
+			referenceType && referenceType->dataStoredIn(DataLocation::CallData)
+			)
+		{
+			// call-data
+			solAssert(referenceType->isImplicitlyConvertibleTo(*TypeProvider::bytesCalldata()), "");
+			IRVariable var = convert(*arguments[0], *TypeProvider::bytesCalldata());
+			templ("abiDecode", m_context.abiFunctions().tupleDecoder(targetTypes, false));
+			templ("offset", var.part("offset").name());
+			templ("length", var.part("length").name());
+//			templ("length", "calldatasize()");
+//			utils().convertType(*referenceType, *TypeProvider::bytesCalldata());
+//			utils().abiDecode(targetTypes, false);
+		}
+		else
+		{
+//			solUnimplementedAssert(false, "");
+			// memory
+			IRVariable var = convert(*arguments[0], *TypeProvider::bytesMemory());
+//			std::cout << "- from memory: " << var.type().toString() << " @ " << var.type().memoryHeadSize() << std::endl;
+//			std::cout << "- from memory: " << firstArgType->toString() << " @ " << firstArgType->memoryHeadSize() << std::endl;
+//			std::cout << "- from memory: " << referenceType->toString() << " @ " << referenceType->memoryHeadSize() << std::endl;
+			for (auto& slot : var.stackSlots())
+				std::cout << " -> " << slot << std::endl;
```

Comment:
Look at `ArrayType::makeStackItems` to check the stack slots

Reply:
nice 👍